README.md
# aiothrottle
Throttled flow controlling StreamReader for aiohttp

#Usage
```python
import functools
import aiohttp
import aiothrottle

# setup the rate limit
kbps = 200
partial = functools.partial(
    aiothrottle.ThrottledStreamReader, rate_limit=kbps * 1024)
aiohttp.client_reqrep.ClientResponse.flow_control_class = partial

# download a large file without blocking bandwidth
response = aiohttp.request("GET", "http://example.com/largefile.zip")
with open("largefile.zip", "wb") as file:
    read_next = True
    while read_next:
        # read 1 MB chunks
        chunk = response.content.read(2**20)
        file.write(chunk)
        read_next = len(chunk) != 0
response.close()

# unset the rate limit
aiohttp.client_reqrep.ClientResponse.flow_control_class = (
    aiohttp.streams.FlowControlStreamReader)
```


CHANGES.txt
CHANGES
=======

0.1.3 (30-08-2016)
------------------

- Set minimum required Python version to 3.4.2, the same as aiohttp 1.0.0

- Made aiothrottle compatible with current aiohttp version again
  (Now using aiohttp.StreamReader._buffer_size)

- Catching RuntimeError when trying to pause a closing _SSLProtocolTransport

0.1.2 (08-08-2015)
------------------

- Fixed resuming transport too often

- Added 'rate_limit' and 'throttling' properties

- Fixed buffer limit control

0.1.1 (08-02-2015)
------------------

- Added limit_rate() and unlimit_rate() globally and response-wise

- Raising ValueError on invalid rate limit

- Cancelling _check_handle in Throttle's destructor

0.1.0 (08-01-2015)
------------------

- Initial release with basic throttling functionality


tests/test_throttle.py
import asyncio
from unittest import TestCase
from unittest.mock import Mock, patch
import sys
import aiothrottle

PY34 = sys.version_info >= (3, 4)


@asyncio.coroutine
def sleep_mock(*_):
    pass


class TestThrottle(TestCase):

    def setUp(self):
        self.loop = asyncio.new_event_loop()
        asyncio.set_event_loop(None)

    def _make_one(self):
        return aiothrottle.Throttle(limit=10, loop=self.loop)

    def _set_time(self, time):
        return patch.object(self.loop, "time", return_value=time)

    def test_parameters(self):
        with self._set_time(111):
            with patch("asyncio.get_event_loop", return_value=self.loop):
                t = aiothrottle.Throttle(10)
                self.assertIs(t._loop, self.loop)
            t = self._make_one()
        self.assertEqual(t._limit, 10)
        self.assertEqual(t.limit, 10)
        self.assertEqual(t._io, 0)
        self.assertIs(t._loop, self.loop)
        self.assertEqual(t._reset_time, 111)

    def test_invalid_limit(self):
        self.assertRaises(
            ValueError, aiothrottle.Throttle, limit=0)
        self.assertRaises(
            ValueError, aiothrottle.Throttle, limit=-10)
        t = self._make_one()
        self.assertRaises(
            ValueError, setattr, t, "limit", 0)
        self.assertRaises(
            ValueError, setattr, t, "limit", -10)

    def test_time_left(self):
        t = self._make_one()
        t.add_io(2)
        self.assertEqual(t.time_left(), 2/10)

    def test_add_io(self):
        t = self._make_one()
        t.add_io(2)
        self.assertEqual(t._io, 2)

    def test_reset_io(self):
        t = self._make_one()
        t.add_io(2)
        t.reset_io()
        self.assertEqual(t._io, 0)

    @patch("asyncio.sleep", Mock(wraps=sleep_mock))
    def test_wait_remaining(self):
        t = self._make_one()
        t.add_io(2)
        self.loop.run_until_complete(t.wait_remaining())
        asyncio.sleep.assert_called_with(2/10)

    def test_current_rate(self):
        with self._set_time(111):
            t = self._make_one()
            t.add_io(2)
            self.assertRaises(RuntimeError, t.current_rate)

        with self._set_time(116):
            self.assertEqual(t.current_rate(), 2/5)
            t.reset_io()
            self.assertEqual(t.current_rate(), 0)

    def test_within_limit(self):
        with self._set_time(111):
            t = self._make_one()
        t.add_io(2)
        with self._set_time(116):
            self.assertTrue(t.within_limit())
            t.add_io(60)
            self.assertFalse(t.within_limit())


tests/test_with_server.py

import asyncio
import logging
import aiohttp
import functools
import sys
import aiothrottle


@asyncio.coroutine
def run_throttle_test(loop, url):
    res = yield from aiohttp.request("GET", url)
    logging.debug("[test] got response: %s", res)

    start_time = loop.time()

    read_next = True
    amount = 0
    while read_next:
        # read 1 MB chunks
        data = yield from res.content.read(2**20)
        data_len = len(data)
        amount += data_len
        read_next = data_len != 0
    res.close()
    logging.debug("[test] read %d bytes", amount)

    end_time = loop.time()
    if start_time == end_time:
        logging.debug("[test] no time passed!")
    else:
        time_passed = loop.time() - start_time
        bps = amount / time_passed
        logging.info(
            "[test] %d:%02d passed, reading rate: %.3f KB/s",
            time_passed / 60, time_passed % 60, bps / 1024)


@asyncio.coroutine
def run_limit_test(loop, url, kbps=0):
    if kbps > 0:

        logging.info("[test] limiting to %d KB/s", kbps)
        aiothrottle.limit_rate(kbps * 1024)
        yield from run_throttle_test(loop, url)

    else:

        logging.info("[test] unlimiting rate")
        aiothrottle.unlimit_rate()
        yield from run_throttle_test(loop, url)


def main():
    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s.%(msecs)03d - %(levelname)s - %(message)s",
        datefmt="%H:%M:%S",
        stream=sys.stdout,
        # filename=r"R:\output.txt"
        )

    print("started")
    loop = asyncio.get_event_loop()
    url = "http://ipv4.download.thinkbroadband.com/10MB.zip"
    try:
        for kbps in (0,) + tuple(range(100, 850, 50)):
            loop.run_until_complete(run_limit_test(loop, url, kbps))
    except KeyboardInterrupt:
        pass
    finally:
        loop.close()
        print("ended")


main()


tests/test_throttled_stream_reader.py
import asyncio
import functools
from unittest import TestCase
from unittest.mock import Mock, patch
import aiohttp
from aiohttp.client_reqrep import ClientResponse
import aiothrottle


@asyncio.coroutine
def sleep_mock(*_):
    pass


class TestThrottledStreamReader(TestCase):

    def setUp(self):
        self.stream = Mock()
        self.stream.paused = True
        self.stream.transport.is_closing.side_effect = [False]
        self.transp = self.stream.transport
        self.loop = asyncio.new_event_loop()
        asyncio.set_event_loop(None)

    def tearDown(self):
        self.loop.close()

    def _make_one(self):
        r = aiothrottle.ThrottledStreamReader(
            self.stream, rate_limit=10, buffer_limit=10, loop=self.loop)
        return r

    def _set_time(self, time):
        return patch.object(self.loop, "time", return_value=time)

    def test_parameters(self):
        with patch("asyncio.get_event_loop", return_value=self.loop):
            r = aiothrottle.ThrottledStreamReader(self.stream, 10, 1)
            self.assertIs(r._loop, self.loop)

        r = self._make_one()
        self.assertIs(r._loop, self.loop)
        self.assertIsInstance(r._throttle, aiothrottle.Throttle)
        self.assertEqual(r.rate_limit, 10)
        self.assertEqual(r._throttle.limit, 10)
        self.assertIs(r._stream, self.stream)
        self.assertEqual(r._b_limit, 2 * 10)
        self.assertFalse(r._b_limit_reached)
        self.assertIsNone(r._check_handle)
        self.assertTrue(r.throttling)
        self.assertTrue(self.transp.resume_reading.called)

    def test_attribute_error(self):
        with patch.object(
                self, "stream", Mock(spec=["paused"])):
            # test catching AttributeError on stream.transport
            r = self._make_one()
            self.stream.paused = False
            r._try_pause()
            self.stream.paused = True
            r._try_resume()

    def test_limit_rate(self):
        r = self._make_one()
        r.limit_rate(20)
        self.assertEqual(r._throttle.limit, 20)
        self.assertTrue(r.throttling)

    def test_unlimit_rate(self):
        r = self._make_one()
        r.unlimit_rate()
        self.assertFalse(r.throttling)
        self.assertFalse(self.stream.paused)

    def test_global_limit_rate(self):
        aiothrottle.limit_rate(10)
        klass = ClientResponse.flow_control_class
        self.assertIsInstance(klass, functools.partial)
        r = klass(self.stream, loop=self.loop)
        self.assertIsInstance(r, aiothrottle.ThrottledStreamReader)
        self.assertEqual(r._throttle.limit, 10)

    def test_global_unlimit_rate(self):
        aiothrottle.limit_rate(10)
        aiothrottle.unlimit_rate()
        klass = ClientResponse.flow_control_class
        self.assertIs(klass, aiohttp.streams.FlowControlStreamReader)

    def test_pause(self):
        r = self._make_one()
        self.transp.reset_mock()

        self.stream.paused = True
        r._try_pause()
        self.assertFalse(self.transp.pause_reading.called)
        self.assertFalse(self.transp.resume_reading.called)

        self.stream.paused = False
        r._try_pause()
        self.assertTrue(self.transp.pause_reading.called)
        self.assertFalse(self.transp.resume_reading.called)

    def test_resume(self):
        r = self._make_one()
        self.transp.reset_mock()

        self.stream.paused = False
        r._try_resume()
        self.assertFalse(self.transp.pause_reading.called)
        self.assertFalse(self.transp.resume_reading.called)

        self.stream.paused = True
        r._try_resume()
        self.assertFalse(self.transp.pause_reading.called)
        self.assertTrue(self.transp.resume_reading.called)

    @patch("asyncio.sleep", Mock(wraps=sleep_mock))
    def test_read(self):
        r = self._make_one()
        r.feed_data(b'da')
        res = self.loop.run_until_complete(r.read(1))
        self.assertEqual(res, b'd')

    @patch("asyncio.sleep", Mock(wraps=sleep_mock))
    def test_readline(self):
        r = self._make_one()
        r.feed_data(b'data\n')
        res = self.loop.run_until_complete(r.readline())
        self.assertEqual(res, b'data\n')

    @patch("asyncio.sleep", Mock(wraps=sleep_mock))
    def test_readany(self):
        r = self._make_one()
        r.feed_data(b'data')
        res = self.loop.run_until_complete(r.readany())
        self.assertEqual(res, b'data')

    @patch("asyncio.sleep", Mock(wraps=sleep_mock))
    def test_readexactly(self):
        r = self._make_one()
        r.feed_data(b'datadata')
        res = self.loop.run_until_complete(r.readexactly(2))
        self.assertEqual(res, b'da')

    def _make_one_nonfull_buffer(self):
        with self._set_time(111):
            r = self._make_one()
            self.transp.reset_mock()
            r.feed_data(b"data" * 3)
        return r

    def _make_one_full_buffer(self):
        r = self._make_one_nonfull_buffer()
        with self._set_time(111):
            r.feed_data(b"data" * 3)
        return r

    def test_recheck_on_error(self):
        current_time = 111

        def time_mock():
            return current_time

        @asyncio.coroutine
        def timing_sleep_mock(delay, *_):
            # simulate the passing of time
            nonlocal current_time
            current_time += delay

        with patch("asyncio.sleep", Mock(wraps=timing_sleep_mock)):

            with patch.object(self.loop, "time", time_mock):
                r = self._make_one()
                r.feed_data(b"data")
                self.loop.run_until_complete(r.read(4))

    def test_nonfull_buffer(self):
        r = self._make_one_nonfull_buffer()
        self.assertFalse(r._b_limit_reached)

    def test_full_buffer(self):
        r = self._make_one_full_buffer()
        self.assertTrue(r._b_limit_reached)

    def test_scheduling_resume(self):
        r = self._make_one_nonfull_buffer()
        self.assertIsNotNone(r._check_handle)
        self.assertEqual(r._check_handle._when, 111 + 12 / 10)
        mock_handle = Mock()
        r._check_handle = mock_handle
        r._schedule_resume()
        self.assertTrue(mock_handle.cancel.called)

    def test_nonscheduling_resume(self):
        r = self._make_one_full_buffer()
        self.assertIsNone(r._check_handle)

    def test_check_callback(self):
        r = self._make_one_nonfull_buffer()
        r._check_callback()
        self.assertIsNone(r._check_handle)
        self.assertFalse(self.stream.paused)

    def test_callback_cancel(self):
        r = self._make_one()
        handle = Mock()
        r._check_handle = handle
        del r
        self.assertTrue(handle.cancel.called)

    def test_throttling_nonfull_buffer_pausing(self):
        r = self._make_one_nonfull_buffer()
        self.assertTrue(self.stream.paused)

        with self._set_time(112):
            self.loop.run_until_complete(r.read(4))
            self.assertTrue(self.stream.paused)

    def test_throttling_full_buffer_pausing(self):
        r = self._make_one_full_buffer()
        self.assertTrue(self.stream.paused)

        with self._set_time(112):
            self.loop.run_until_complete(r.read(4))
            self.assertTrue(self.stream.paused)

    def test_nonthrottling_nonfull_buffer_nonpausing_read(self):
        r = self._make_one_nonfull_buffer()
        r.unlimit_rate()
        self.assertFalse(self.stream.paused)

        with self._set_time(112):
            self.loop.run_until_complete(r.read(4))
            self.assertFalse(self.stream.paused)

    def test_nonthrottling_full_buffer_pausing_read(self):
        r = self._make_one_full_buffer()
        r.unlimit_rate()
        self.assertTrue(self.stream.paused)

        with self._set_time(112):
            self.loop.run_until_complete(r.read(4))
            self.assertTrue(self.stream.paused)
            self.loop.run_until_complete(r.read(20))
            self.assertFalse(self.stream.paused)

    def test_nonthrottling_nonfull_buffer_nonpausing_feed(self):
        r = self._make_one_nonfull_buffer()
        r.unlimit_rate()
        r.feed_data(b"data")
        self.assertFalse(self.stream.paused)

    def test_nonthrottling_full_buffer_pausing_feed(self):
        r = self._make_one_full_buffer()
        r.unlimit_rate()
        r.feed_data(b"data")
        self.assertTrue(self.stream.paused)

    def test_nonthrottling_nonfull_buffer_resuming_feed(self):
        r = self._make_one_nonfull_buffer()
        r.unlimit_rate()
        r._try_pause()
        r.feed_data(b"data")
        self.assertFalse(self.stream.paused)

    def test_nonthrottling_full_buffer_nonresuming_feed(self):
        r = self._make_one_full_buffer()
        r.unlimit_rate()
        r._try_pause()
        r.feed_data(b"data")
        self.assertTrue(self.stream.paused)


__main__.py
#!/usr/bin/env python3

import asyncio
import aiohttp
import logging
from .throttle import ThrottledFlowControlStreamReader


class TestReadTransport(asyncio.ReadTransport):
    def __init__(
            self, protocol,
            total_size, chunk_size, interval=1.0):
        super().__init__()
        self._protocol = protocol
        self._total_size = total_size
        self._chunk_size = int(chunk_size)
        self._interval = interval
        self._paused = False
        self._closed = False
        self._loop = asyncio.get_event_loop()
        self._feed_handle = None
        self._bytes_fed = 0
        self.closed_callback = None

    def _schedule_data_feeding(self):
        if self._feed_handle is not None:
            self._feed_handle.cancel()
        self._feed_handle = self._loop.call_later(
            self._interval, self._feed_data)
        logging.debug("[transport] scheduled data feed")

    def _feed_data(self):
        if self._closed:
            raise RuntimeError("closed")
        logging.debug("[transport] feeding %d bytes", self._chunk_size)
        self._protocol.data_received(bytes(self._chunk_size))
        self._bytes_fed += self._chunk_size

        if self._bytes_fed >= self._total_size:
            self.close()
        elif not self._paused:
            self._schedule_data_feeding()

    def open(self):
        if self._closed:
            raise RuntimeError("closed")
        self._protocol.connection_made(self)
        logging.debug("[transport] opened")
        self._feed_data()

    def pause_reading(self):
        if self._closed:
            raise RuntimeError("closed")
        if self._paused:
            raise RuntimeError("already paused")
        if self._feed_handle is not None:
            self._feed_handle.cancel()
            self._feed_handle = None
        self._paused = True
        logging.debug("[transport] paused")

    def resume_reading(self):
        if self._closed:
            raise RuntimeError("closed")
        if not self._paused:
            raise RuntimeError("not paused")
        self._schedule_data_feeding()
        self._paused = False
        logging.debug("[transport] resumed")

    def close(self):
        if self._closed:
            return
        self._protocol.eof_received()
        self._protocol.connection_lost(None)
        self.pause_reading()
        self._closed = True
        if self.closed_callback is not None:
            self._loop.call_soon(self.closed_callback)
        logging.debug("[transport] closed")


@asyncio.coroutine
def run_reader_test():
    # test transport: write 1024 bytes, 100 bytes per second
    # throttled reader: limit transmit rate to 40 bytes per second,
    #  while data is requested in 200 byte chunks

    closed_waiter = asyncio.Future()

    def transport_closed():
        logging.debug("[test] got transport closed callback")
        closed_waiter.set_result(None)

    stream = aiohttp.StreamParser()
    reader = ThrottledFlowControlStreamReader(stream, rate_limit=40)

    protocol = asyncio.StreamReaderProtocol(stream)

    transport = TestReadTransport(
        protocol, total_size=1024, chunk_size=100)
    transport.closed_callback = transport_closed
    transport.open()

    read_more = True
    attempt = 0
    amount = 0
    loop = asyncio.get_event_loop()
    start_time = loop.time()
    while read_more:
        data = yield from reader.read(200)
        data_len = len(data)

        logging.debug(
            "[test] read attempt %d: read %d bytes",
            attempt, data_len)
        attempt += 1
        amount += data_len

        read_more = data_len != 0

    logging.info(
        "[test] reading rate: %.3f bytes per second",
        amount / (loop.time() - start_time))

    yield from closed_waiter


def main():
    logging.basicConfig(
        level=logging.DEBUG,
        format="%(asctime)s - %(levelname)s - %(message)s",
        datefmt="%H:%M:%S")
    loop = asyncio.get_event_loop()

    print("started")
    try:
        loop.run_until_complete(run_reader_test())
    except KeyboardInterrupt:
        pass
    finally:
        loop.close()
        print("ended")

main()


examples/throttled_stream_reader.py
#!/usr/bin/env python3

import asyncio
import os
import aiohttp
import aiothrottle


@asyncio.coroutine
def load_file(url, loop):
    response = yield from aiohttp.request("GET", url, loop=loop)
    size = int(response.headers.get("Content-Length", "0"))

    start_time = loop.time()

    with open(os.devnull, "wb") as file:
        read_next = True
        loaded = 0
        prev_progress = -1
        prev_print_time = start_time
        prev_print_loaded = 0
        while read_next:
            # read 1 MB chunks
            chunk = yield from response.content.read(2**20)
            file.write(chunk)

            chunk_len = len(chunk)
            loaded += chunk_len

            progress = int(loaded * 100 / size)
            if progress and progress != prev_progress:
                # print progress and current download rate
                now = loop.time()
                period = now - prev_print_time
                loaded_since = loaded - prev_print_loaded
                temp_rate = loaded_since / period
                print("%d%% - %d KB/s" % (progress, temp_rate / 1024))
                prev_print_time = now
                prev_print_loaded = loaded
            prev_progress = progress

            read_next = chunk_len != 0
    response.close()

    end_time = loop.time()
    download_time = end_time - start_time

    print(
        "download rate: %d KB/s" %
        int(size / download_time / 1024))


def main():
    # setup the rate limit to 200 KB/s
    aiothrottle.limit_rate(200 * 1024)

    # download a large file without blocking bandwidth
    loop = asyncio.get_event_loop()
    loop.run_until_complete(load_file(
        "http://ipv4.download.thinkbroadband.com/10MB.zip", loop))

    # unset the rate limit
    aiothrottle.unlimit_rate()

if __name__ == '__main__':
    try:
        main()
    except KeyboardInterrupt:
        pass


aiothrottle.py
#!/usr/bin/env python3

import asyncio
import aiohttp
import logging

class Throttle:
    """Throttle for IO operations"""

    def __init__(self, rate_limit, interval=1.0, loop=None):
        """
        :param rate_limit: the limit in bytes to read/write per interval
        :type: int
        :param interval: the limitation time frame in seconds
        :type: float
        """
        self.rate_limit = rate_limit
        self.interval = interval
        self._interv_start = 0
        self._io_in_interv = 0
        if loop is None:
            loop = asyncio.get_event_loop()
        self._loop = loop

    def _reset_interval(self, timestamp=-1):
        """starts a new interval at the given timestamp

        :param timestamp: the timestamp in seconds of the interval start
        :type: float
        :rtype: None
        """
        self._io_in_interv = 0
        self._interv_start = timestamp if timestamp >= 0 else self._loop.time()
        logging.debug("[throttle] reset interval")

    def time_left(self):
        """returns the number of seconds left in the current interval

        If the interval has already passed, this returns 0.

        :returns: seconds left until the interval ends
        :rtype: float
        """
        now = self._loop.time()
        if now - self._interv_start > self.interval:
            logging.debug("[throttle] seconds left: 0")
            return 0

        remaining = self.interval - (now - self._interv_start)
        logging.debug("[throttle] seconds left: %.3f", remaining)
        return remaining

    def allowed_io(self):
        """checks if a requested IO action is allowed

        :returns: number of bytes allowed to read/write
        :rtype: int
        """
        if self.time_left() == 0:
            self._reset_interval()
        allowed = self.rate_limit - self._io_in_interv
        logging.debug("[throttle] allowed bytes: %d", allowed)
        return allowed

    def add_io(self, byte_count):
        """registers a number of bytes read/written

        :param byte_count: number of bytes to add to the current interval
        :type: int
        :rtype: None
        """
        self._io_in_interv += byte_count
        logging.debug(
            "[throttle] added bytes: %d, now: %d",
            byte_count, self._io_in_interv)

    @asyncio.coroutine
    def wait_remaining(self):
        """waits until the current interval has passed

        This makes sure a new IO action is allowed
        :rtype: None
        """
        time_left = self.time_left()
        logging.debug("[throttle] sleeping for %.3f seconds", time_left)
        yield from asyncio.sleep(time_left)

class ThrottledFlowControlStreamReader(aiohttp.StreamReader):

    def __init__(
            self, stream, rate_limit, interval=1.0,
            buffer_limit=2**16, loop=None, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self._loop = loop or asyncio.get_event_loop()
        self._throttle = Throttle(
            rate_limit, interval, self._loop)
        self._stream = stream
        self._b_limit = buffer_limit * 2

        # resume transport reading
        if stream.paused:
            try:
                self._stream.transport.resume_reading()
            except AttributeError:
                pass

    def _try_pause(self):
        try:
            self._stream.transport.pause_reading()
        except AttributeError:
            pass
        else:
            self._stream.paused = True

    def _try_resume(self):
        try:
            self._stream.transport.resume_reading()
        except AttributeError:
            pass
        else:
            self._stream.paused = False

    def feed_data(self, data, size=0):
        has_waiter = (
            self._waiter is not None and not self._waiter.cancelled())

        super().feed_data(data)

        if (
                not self._stream.paused and
                not has_waiter and len(self._buffer) > self._b_limit):
            self._try_pause()

    def _maybe_resume(self):
        size = len(self._buffer)
        if self._stream.paused:
            if size < self._b_limit:
                self._try_resume()
        else:
            if size > self._b_limit:
                self._try_pause()

    @asyncio.coroutine
    def read(self, byte_count=-1):
        logging.debug("reading %d bytes", byte_count)
        data = yield from super().read(byte_count)
        self._maybe_resume()
        return data

    @asyncio.coroutine
    def readline(self):
        logging.debug("reading line")
        data = yield from super().readline()
        self._maybe_resume()
        return data

    @asyncio.coroutine
    def readany(self):
        logging.debug("reading anything")
        data = yield from super().readany()
        self._maybe_resume()
        return data

    @asyncio.coroutine
    def readexactly(self, byte_count):
        logging.debug("reading exactly %d bytes", byte_count)
        data = yield from super().readexactly(byte_count)
        self._maybe_resume()
        return data

class TestReadTransport(asyncio.ReadTransport):
    def __init__(
            self, protocol,
            total_size, chunk_size, interval=1.0):
        super().__init__()
        self._protocol = protocol
        self._total_size = total_size
        self._chunk_size = int(chunk_size)
        self._interval = interval
        self._paused = False
        self._closed = False
        self._loop = asyncio.get_event_loop()
        self._feed_handle = None
        self._bytes_fed = 0
        self.closed_callback = None

    def _schedule_data_feeding(self):
        if self._feed_handle is not None:
            self._feed_handle.cancel()
        self._feed_handle = self._loop.call_later(
            self._interval, self._feed_data)
        logging.debug("[transport] scheduled data feed")

    def _feed_data(self):
        if self._closed:
            raise RuntimeError("closed")
        logging.debug("[transport] feeding %d bytes", self._chunk_size)
        self._protocol.data_received(bytes(self._chunk_size))
        self._bytes_fed += self._chunk_size

        if self._bytes_fed >= self._total_size:
            self.close()
        elif not self._paused:
            self._schedule_data_feeding()

    def open(self):
        if self._closed:
            raise RuntimeError("closed")
        self._protocol.connection_made(self)
        logging.debug("[transport] opened")
        self._feed_data()

    def pause_reading(self):
        if self._closed:
            raise RuntimeError("closed")
        if self._paused:
            raise RuntimeError("already paused")
        if self._feed_handle is not None:
            self._feed_handle.cancel()
            self._feed_handle = None
        self._paused = True
        logging.debug("[transport] paused")

    def resume_reading(self):
        if self._closed:
            raise RuntimeError("closed")
        if not self._paused:
            raise RuntimeError("not paused")
        self._schedule_data_feeding()
        self._paused = False
        logging.debug("[transport] resumed")

    def close(self):
        if self._closed:
            return
        self._protocol.eof_received()
        self._protocol.connection_lost(None)
        self.pause_reading()
        self._closed = True
        if self.closed_callback is not None:
            self._loop.call_soon(self.closed_callback)
        logging.debug("[transport] closed")

@asyncio.coroutine
def run_reader_test():
    # test transport: write 1024 bytes, 100 bytes per second
    # throttled reader: limit transmit rate to 40 bytes per second,
    #  while data is requested in 200 byte chunks

    closed_waiter = asyncio.Future()

    def transport_closed():
        logging.debug("[test] got transport closed callback")
        closed_waiter.set_result(None)

    stream = aiohttp.StreamParser()
    reader = ThrottledFlowControlStreamReader(stream, rate_limit=40)

    protocol = asyncio.StreamReaderProtocol(stream)

    transport = TestReadTransport(
        protocol, total_size=1024, chunk_size=100)
    transport.closed_callback = transport_closed
    transport.open()

    read_more = True
    attempt = 0
    amount = 0
    loop = asyncio.get_event_loop()
    start_time = loop.time()
    while read_more:
        data = yield from reader.read(200)
        data_len = len(data)

        logging.debug(
            "[test] read attempt %d: read %d bytes",
            attempt, data_len)
        attempt += 1
        amount += data_len

        read_more = data_len != 0

    logging.info(
        "[test] reading rate: %.3f bytes per second",
        amount / (loop.time() - start_time))

    yield from closed_waiter

if __name__ == "__main__":
    logging.basicConfig(
        level=logging.DEBUG,
        format="%(asctime)s - %(levelname)s - %(message)s",
        datefmt="%H:%M:%S")
    loop = asyncio.get_event_loop()

    print("started")
    try:
        loop.run_until_complete(run_reader_test())
    except KeyboardInterrupt:
        pass
    finally:
        loop.close()
        print("ended")
